### PR TITLE
Create new manifest after deleting obsolete files

### PR DESCRIPTION
**Pull request by v01dstar**
Close https://github.com/tikv/tikv/issues/16256

### CODE REVIEW

File: src/blob_file_set.cc [hunk 1]
```diff
@@ -136,8 +131,8 @@ Status BlobFileSet::Recover() {
                    "Titan recovery delete obsolete file %s.", f.c_str());
     env_->DeleteFile(dirname_ + "/" + f);
   }
-
-  return Status::OK();
+  auto new_manifest_file_number = NewFileNumber();
+  return OpenManifest(new_manifest_file_number);
```
Comment (Connor1996) [inline src/blob_file_set.cc:135]: do not change the order, otherwise the old manifest would be already deleted when failed to create new manifest

Comment (v01dstar) [reply src/blob_file_set.cc]: PTAL

File: src/blob_file_set.cc [hunk 2]
```diff
@@ -129,18 +127,17 @@ Status BlobFileSet::Recover() {
     FileType file_type;
     if (!ParseFileName(f, &file_number, &file_type)) continue;
     if (alive_files.find(file_number) != alive_files.end()) continue;
+    // The newly created manifest file is not obsolete.
```
Comment (Connor1996) [inline src/blob_file_set.cc:130]: Please describe why need this check

**Review by Connor1996 — COMMENTED** [review]
Please cover it by test

File: src/blob_file_set.cc [hunk 3]
```diff
@@ -129,18 +127,17 @@ Status BlobFileSet::Recover() {
...
       continue;
     TITAN_LOG_INFO(db_options_.info_log,
                    "Titan recovery delete obsolete file %s.", f.c_str());
     env_->DeleteFile(dirname_ + "/" + f);
   }
-
   return Status::OK();
-}
-
-Status BlobFileSet::OpenManifest(uint64_t file_number) {
```
Comment (tonyxuqqi) [inline src/blob_file_set.cc:143]: is it the typo?

**Comment by tonyxuqqi** [comment]
I don't get the overall idea. To me it seems the logic has no difference. 
Before the change, it will only check 
`"if (alive_files.find(file_number) != alive_files.end()) continue;"`
The original code makes new_manifest_file_number part of alive_files.

But now, alive_files do not ocntain new_manifest_file_number, but the code check it and skip explicitly. 
` if (file_number == new_manifest_file_number && file_type == kDescriptorFile)`

**Comment by v01dstar** [comment]
> I don't get the overall idea. To me it seems the logic has no difference. Before the change, it will only check `"if (alive_files.find(file_number) != alive_files.end()) continue;"` The original code makes new_manifest_file_number part of alive_files.
> 
> But now, alive_files do not ocntain new_manifest_file_number, but the code check it and skip explicitly. ` if (file_number == new_manifest_file_number && file_type == kDescriptorFile)`

It was a orphaned blob file that can not be GCed since it has the same file number as the manifest (they have different prefix and suffix though).

File: src/blob_file_set.cc (hunk 3 above)
Comment (v01dstar) [reply src/blob_file_set.cc]: Yes, fixed.

File: src/blob_file_set.cc (hunk 2 above)
Comment (v01dstar) [reply src/blob_file_set.cc]: Done

**Review by tonyxuqqi — APPROVED** [review]

File: src/blob_file_set.cc [hunk 4]
```diff
@@ -129,14 +127,16 @@ Status BlobFileSet::Recover() {
     FileType file_type;
     if (!ParseFileName(f, &file_number, &file_type)) continue;
     if (alive_files.find(file_number) != alive_files.end()) continue;
+    // We will delete the old manifest file and keep the new one.
```
Comment (Connor1996) [inline src/blob_file_set.cc:130]:
```suggestion
    // Some GC output files may not recorded in the manifest for the last time, so there may be a blob file of as same file number as the new manifest
```

Comment (v01dstar) [reply src/blob_file_set.cc]: PTAL

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**